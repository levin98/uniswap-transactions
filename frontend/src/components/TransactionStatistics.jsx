import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Card, Col, Row, Statistic, Alert } from 'antd';

const TransactionStatistics = ({ transactionList }) => {
    const [error, setError] = useState('');
    const [price, setPrice] = useState(0);

    const getTotalFeeinETH = () => {
        return transactionList.reduce((acc, curr) => {
            return acc + (parseFloat(curr.gas_used) * parseFloat(curr.gas_price)) * Math.pow(10, -18);
        }, 0)
    }

    const getTickerPrice = async () => {
        setError('');
        try {
            const response = await fetch("https://data-api.binance.vision/api/v3/ticker/price?symbol=ETHUSDT");
            const data = await response.json();
            setPrice(data.price);
        } catch (e) {
            setError(e.message);
        }
    }

    useEffect(() => {
        let id = setTimeout(getTickerPrice, 1000);
        return () => clearTimeout(id);
    });

    return (
        <>
            {error && <Alert message={error} type="error" showIcon style={{ marginBottom: '10px' }} />}
            <Row gutter={16}>
                <Col span={8}>
                    <Card><Statistic title="Total Transaction Fee (USDT)" value={getTotalFeeinETH() * price} /></Card>
                </Col>
                <Col span={8}>
                    <Card><Statistic title="Total Transaction Fee (ETH)" value={getTotalFeeinETH()} /></Card>
                </Col>
                <Col span={8}>
                    <Card><Statistic title="ETH/USDT" value={price} /></Card>
                </Col>
            </Row>
        </>
    )
};

TransactionStatistics.propTypes = {
    transactionList: PropTypes.array
}

TransactionStatistics.defaultProps = {
    transactionList: []
}

export default TransactionStatistics;