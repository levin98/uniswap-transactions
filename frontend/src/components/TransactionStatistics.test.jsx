import React from 'react';
import { render, screen } from '@testing-library/react';
import TransactionStatistics from './TransactionStatistics';
import "../mock"

jest.spyOn(window, "fetch").mockImplementation(() => ({ "price": 1000 }));

const mockData = [
    {
        "gasPrice": "19156218168",
        "gasUsed": "422092",
    },
    {
        "gasPrice": "19992909881",
        "gasUsed": "210998",
    }
]

test('renders statistics', () => {
    const { container } = render(<TransactionStatistics transactionList={mockData} />);

    expect(screen.getByText("Total Transaction Fee (USDT)")).toBeInTheDocument();
    expect(screen.getByText("Total Transaction Fee (ETH)")).toBeInTheDocument();
    expect(screen.getByText("ETH/USDT")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
});
