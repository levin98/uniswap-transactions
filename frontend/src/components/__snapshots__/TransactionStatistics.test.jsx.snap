// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders statistics 1`] = `
<div>
  <div
    class="ant-row css-dev-only-do-not-override-k7429z"
    style="margin-left: -8px; margin-right: -8px;"
  >
    <div
      class="ant-col ant-col-8 css-dev-only-do-not-override-k7429z"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-card ant-card-bordered css-dev-only-do-not-override-k7429z"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="ant-statistic css-dev-only-do-not-override-k7429z"
          >
            <div
              class="ant-statistic-title"
            >
              Total Transaction Fee (USDT)
            </div>
            <div
              class="ant-statistic-content"
            >
              <span
                class="ant-statistic-content-value"
              >
                NaN
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-8 css-dev-only-do-not-override-k7429z"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-card ant-card-bordered css-dev-only-do-not-override-k7429z"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="ant-statistic css-dev-only-do-not-override-k7429z"
          >
            <div
              class="ant-statistic-title"
            >
              Total Transaction Fee (ETH)
            </div>
            <div
              class="ant-statistic-content"
            >
              <span
                class="ant-statistic-content-value"
              >
                NaN
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-8 css-dev-only-do-not-override-k7429z"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-card ant-card-bordered css-dev-only-do-not-override-k7429z"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="ant-statistic css-dev-only-do-not-override-k7429z"
          >
            <div
              class="ant-statistic-title"
            >
              ETH/USDT
            </div>
            <div
              class="ant-statistic-content"
            >
              <span
                class="ant-statistic-content-value"
              >
                <span
                  class="ant-statistic-content-value-int"
                >
                  0
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
